mergin/report.py

```diff
@@ -0,0 +1,237 @@
+    elif envelope_byte == 2:
+        envelope_size = 48
+    elif envelope_byte == 3:
+        envelope_size = 48
+    elif envelope_byte == 4:
+        envelope_size = 64
+
+    return no_envelope_header_size + envelope_size
+
+
+class ChangesetReportEntry:
+    """ Derivative of geodiff ChangesetEntry suitable for further processing/reporting """
+    def __init__(self, changeset_entry, geom_idx, geom):
+        self.table = changeset_entry.table.name
+        self.geom_type = geom["type"]
+        self.crs = geom["srs_id"]
+
+        if changeset_entry.operation == changeset_entry.OP_DELETE:
+            self.operation = "delete"
+            self.old_geom = changeset_entry.old_values[geom_idx]
+            self.new_geom = None
+        elif changeset_entry.operation == changeset_entry.OP_UPDATE:
+            self.operation = "update"
+            self.old_geom = changeset_entry.old_values[geom_idx]
+            self.new_geom = changeset_entry.new_values[geom_idx]
+        elif changeset_entry.operation == changeset_entry.OP_INSERT:
+            self.operation = "insert"
+            self.old_geom = None
+            self.new_geom = changeset_entry.new_values[geom_idx]
+
+        self.count = None
+        self.length = None
+        self.area = None
+
+        if self.geom_type == "LINESTRING":
+            # we calculate change in length, for attributes changes only we set to 0
+            self.metric = "length"
+            if self.operation == "delete":
+                self.length = self.measure(self.old_geom)
+            elif self.operation == "update":
+                self.length = self.measure(self.new_geom) - self.measure(self.old_geom)
+            elif self.operation == "insert":
+                self.length = self.measure(self.new_geom)
+        elif self.geom_type == "POLYGON":
+            # we calculate change in area, for attributes changes only we set to 0
+            self.metric = "area"
+            if self.operation == "delete":
+                self.area = self.measure(self.old_geom)
+            elif self.operation == "update":
+                self.area = self.measure(self.new_geom) - self.measure(self.old_geom)
+            elif self.operation == "insert":
+                self.area = self.measure(self.new_geom)
+        else:
+            # regardless of geometry change count as 1
+            self.metric = "count"
+            self.count = 1
+
+    def measure(self, geom):
+        """ Return length or area of geometry based on type """
+        # calculate geom length/area only if QGIS API is available
+        try:
+            from qgis.core import QgsGeometry, QgsDistanceArea, QgsCoordinateReferenceSystem, QgsCoordinateTransformContext
+        except ImportError:
+            return -1
+
+        d = QgsDistanceArea()
+        d.setEllipsoid('WGS84')
+        crs = QgsCoordinateReferenceSystem()
+        crs.createFromString(self.crs)
+        d.setSourceCrs(crs, QgsCoordinateTransformContext())
```

Comment:
ideally the preparation of QgsDistanceArea should be done once, and then called many times - these init functions may need to do some queries in the projection database etc, which may slow things down a lot